### PR TITLE
fix(yaml): fix captures for sequences that contain mappings

### LINKED ISSUE #7532: Semgrep Findings Range Extends Too Far
**Describe the bug**

I noticed that semgrep against yaml, when finding an item in a list, will match the whitespace up to the next item. And depending on how you interpret the end column (is it the last character affected, or the first character not affected), is matching the `-` that starts the next item.

The first match and on line 21, column, 3, the `-`.

```
  20   │   # ok: metadata-technology
  21   │   - id: subprocess-run
```

The second match ends on the next byte after the end of the file, column 20, but there's only 19 bytes, the last byte being the new-line.

This behaviour causes editors to using the vscode plugin to show right up to the next item as having a diagnostic error.
IMO the whitespace on line 21 belongs to the subprocess-run item, not the preceding item, and so the match should end on line 20.
It would be nice to not include line 20, as it's a comment intended for the next item, but I think its reasonable design decision to include the comment, and a more complicated rule could maybe exclude it, but add complexity to rule-writing.

**To Reproduce**
Steps to reproduce the behavior, ideally a link to https://semgrep.dev:

https://semgrep.dev/playground/s/ben-elttam:metadata-technology?editorMode=advanced

**Expected behavior**

For at least YAML, I'd say the whitespace before a list item belongs to the list item on that line, not the item preceding it.

It could be common for people/tools to interpret the end column to be the last column affected, rather than the first column not affected. Vim, with CoC, and semgrep-vscode will highlight the `-` in column 3 as affected. 

**Screenshots**

image

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [X] P2: regular bug that should get fixed

**Environment**
If not using semgrep.dev: are you running off docker, an official binary, a local build?

Semgrep v1.17.1
semgrep-vscode heads/master-0-gb1cea96

**Use case**
What will fixing this bug enable for you?

Vim, with CoC, and semgrep-vscode won't highlight the whitespace preceding the item that is not affected.

### LINKED ISSUE #8226: [YAML] Excessive capture when matching sequences
**Describe the bug**
We have a rule similar to the following:
```
rules:
    - id: foo
      languages: [yaml]
      pattern: |
        importantKey:
          ...
```

Which we intend to match blocks like this

```
arbitraryKey:
  importantKey:
    prop1: true
    prop2: true
    prop3: false
```

or

```
arbitraryKey:
  importantKey:
    - something: true
      somethingElse: false
```

However, we noticed that the start and end location for the match can include the whitespace before the next key if `importantKey` is a sequence. Additionally, it include comments even if they're not at the correct level of indentation.

**To Reproduce**
https://semgrep.dev/playground/s/p87d

**Expected behavior**
The newline + whitespace (+ and comments) should not be included.

**Screenshots**
See semgrep playground link

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [x] P2: regular bug that should get fixed

**Environment**
Semgrep.dev and semgrep via pypi

**Use case**
We used `generic` language mode to address this issue, but has resulted in much slower scans with this rule and requiring us to play with the `generic_ellipsis_max_span` option to work for our use-cases.

This bug being fixed would allow us to revert the rule to a `yaml` rule which appears to perform much better.

### PR DESCRIPTION
**Pull request by MichaHoffmann**
fixes https://github.com/returntocorp/semgrep/issues/8226, pretty sure it also fixes https://github.com/returntocorp/semgrep/issues/7532 and https://github.com/returntocorp/semgrep/issues/5607

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by MichaHoffmann** [comment]
It turns out that we didnt adapt the last captured event when we corrected its token position in mappings. That inhibited us from using the same trick for sequences.

before:
```
┌─────────────────┐
│ 4 Code Findings │
└─────────────────┘

    test.yaml
       foo
          example

            3┆ importantKey:
            4┆   prop1: true
            ⋮┆----------------------------------------
            9┆   importantKey:
           10┆     - idk: true
           11┆ # (BAD) this comment is still included in the match
            ⋮┆----------------------------------------
           15┆ importantKey:
           16┆   prop1: true
            ⋮┆----------------------------------------
           21┆ importantKey:
           22┆   - idk: true
           23┆ shouldNotInclude: true
```
after:
```
┌─────────────────┐
│ 4 Code Findings │
└─────────────────┘

    test.yaml
       foo
          example

            3┆ importantKey:
            4┆   prop1: true
            ⋮┆----------------------------------------
            9┆ importantKey:
           10┆   - idk: true
            ⋮┆----------------------------------------
           15┆ importantKey:
           16┆   prop1: true
            ⋮┆----------------------------------------
           21┆ importantKey:
           22┆   - idk: true
```

**Review by emjin — APPROVED** [review]
I'll add a test in another PR

File: src/parsing/yaml_to_generic.ml [hunk 1]
```diff
@@ -348,8 +348,20 @@ let parse (env : env) : G.expr list =
...
-    | E.Sequence_end, pos -> (acc, pos)
+    | E.Sequence_end, pos ->
+        (* The `Sequence_end` event position is inaccurate: it specifies the *)
+        (* position immediately prior to the next token. This causes comments *)
+        (* and the leading whitespace of the next yaml element to be chomped on *)
+        (* autofix. *)
+        let pos' = Option.fold ~none:pos ~some:(fun (_, pos) -> pos) last_ev in
+        (* Important: we mess with the pos, we need to propagate it for consumers *)
+        (* of env.last_ev. *)
+        env.last_event <- Some (E.Mapping_end, pos');
```
Comment (emjin) [inline src/parsing/yaml_to_generic.ml]:
```suggestion
        env.last_event <- Some (E.Sequence_end, pos');
```

**Review by emjin — CHANGES_REQUESTED** [review]
oops sorry one thing that needs to be changed

File: src/parsing/yaml_to_generic.ml (hunk 1 above)
Comment (MichaHoffmann) [reply src/parsing/yaml_to_generic.ml]: oh, yeah good catch

Comment (MichaHoffmann) [reply src/parsing/yaml_to_generic.ml]: done

**Comment by MichaHoffmann** [comment]
failed test worked for me locally, might be flakiness
```
⬢[mhoffm@toolbox cli]$ pipenv run pytest --snapshot-update tests/e2e/test_ls.py::test_ls_multi
============================================================= test session starts =============================================================
platform linux -- Python 3.11.4, pytest-7.3.0, pluggy-1.0.0
rootdir: /var/home/mhoffm/git/semgrep/cli
configfile: pyproject.toml
plugins: freezegun-0.4.2, xdist-3.2.1, split-0.8.1, snapshot-0.9.0, rerunfailures-11.1.2, mock-3.10.0
collected 1 item

tests/e2e/test_ls.py .                                                                                                                  [100%]

============================================================== warnings summary ===============================================================
../../../.local/share/virtualenvs/cli-kEnraMvm/lib64/python3.11/site-packages/pytest_freezegun.py:17
../../../.local/share/virtualenvs/cli-kEnraMvm/lib64/python3.11/site-packages/pytest_freezegun.py:17
  /var/home/mhoffm/.local/share/virtualenvs/cli-kEnraMvm/lib64/python3.11/site-packages/pytest_freezegun.py:17: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(pytest.__version__) < LooseVersion('3.6.0'):

tests/e2e/test_ls.py::test_ls_multi
  /var/home/mhoffm/git/semgrep/cli/src/semgrep/semgrep_core.py:21: DeprecationWarning: path is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
    with importlib.resources.path("semgrep.bin", exec_name) as path:

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================== 1 passed, 3 warnings in 4.66s ========================================================

```

**Review by emjin — APPROVED** [review]
Looks good now!

**Comment by aryx** [comment]
Would be good to add test for this. Maybe the one used in the issues supposidely fixed by this PR.
When it comes to range, there's no easy way to use tests/patterns/ or tests/rules to show that it works. You can use
cli/tests/e2e/ and its snapshot based testing to dump the json for the range and assert they are right.

**Comment by emjin** [comment]
@MichaHoffman I didn't end up getting time to add a test today, so if you're able to that would be great! Otherwise I can merge tomorrow and then follow up with a test

**Comment by MichaHoffmann** [comment]
Added a basic snapshot test with the example from the issue.

**Comment by aryx** [comment]
Need to fix precommit @MichaHoffmann 
need pre-commit install and then precommit run -a

**Comment by MichaHoffmann** [comment]
> Need to fix precommit @MichaHoffmann need pre-commit install and then precommit run -a

fixed the formatting issue

**Comment by aryx** [comment]
@ajbt200128 can you take a look at this PR? Seems like it causes some regressions with test_ls here: https://github.com/returntocorp/semgrep/actions/runs/5691293193/job/15426494310?pr=8388
```

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/e2e/test_ls.py::test_ls_multi - IndexError: list index out of range
= 1 failed, 1385 passed, 1 skipped, 2 xfailed, 5588 warnings in 173.56s (0:02:53) =
Error: Process completed with exit code 1.

```

**Comment by mhoffm-aiven** [comment]
> @ajbt200128 can you take a look at this PR? Seems like it causes some regressions with test_ls here: https://github.com/returntocorp/semgrep/actions/runs/5691293193/job/15426494310?pr=8388
> 
> ```
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> =========================== short test summary info ============================
> FAILED tests/e2e/test_ls.py::test_ls_multi - IndexError: list index out of range
> = 1 failed, 1385 passed, 1 skipped, 2 xfailed, 5588 warnings in 173.56s (0:02:53) =
> Error: Process completed with exit code 1.
> ```

fwiw: that passed locally for me

**Comment by emjin** [comment]
I think develop is having a problem with test_ls

**Comment by ajbt200128** [comment]
> I think develop is having a problem with test_ls

Yea none of these changes effect the LS

**Comment by ajbt200128** [comment]
Lol tests are passing except for benchmarks now. I think github is not having a good day

**Comment by emjin** [comment]
Benchmarks fail because it tries to post a PR comment but doesn't have a token

**Pull request merged**